sdcflows/interfaces/bspline.py

```diff
@@ -0,0 +1,249 @@
+        nb.Nifti1Image(data * mask - interp_data, fmapnii.affine, fmapnii.header).to_filename(
+            self._results["out_error"])
+
+        if not self.inputs.extrapolate:
+            return runtime
+
+        bg_indices = np.argwhere(~mask)
+        bg_points = (
+            oriented_nii.affine.astype("float32") @ (
+                np.vstack((bg_indices.T, np.ones((1, bg_indices.shape[0],), dtype=int)))
+            )
+        )[:3].T
+
+        extrapolators = np.vstack(
+            [bspline_weights(bg_points, level) for level in bs_levels]
+        )
+        interp_data[~mask] = (
+            (np.array(model.coef_) @ extrapolators)  # Extrapolation
+            .astype("float32")
+        )
+        self._results["out_extrapolated"] = out_name.replace("_field.", "_extra.")
+        nb.Nifti1Image(interp_data, fmapnii.affine, hdr).to_filename(
+            self._results["out_extrapolated"]
+        )
+        return runtime
+
+
+def canonical_orientation(img):
+    """Generate an alternative image aligned with the array axes."""
+    if isinstance(img, (str, Path)):
+        img = nb.load(img)
+
+    shape = np.array(img.shape[:3])
+    affine = np.diag(np.hstack((img.header.get_zooms()[:3], 1)))
+    affine[:3, 3] -= affine[:3, :3] @  (0.5 * (shape - 1))
+    nii = nb.Nifti1Image(img.dataobj, affine)
+    nii.header.set_xyzt_units(*img.header.get_xyzt_units())
+    return nii
+
+
+def bspline_grid(img, control_zooms_mm=DEFAULT_ZOOMS_MM):
+    """Calculate a Nifti1Image object encoding the location of control points."""
+    if isinstance(img, (str, Path)):
+        img = nb.load(img)
+
+    im_zooms = np.array(img.header.get_zooms())
+    im_shape = np.array(img.shape[:3])
+
+    # Calculate the direction cosines of the target image
+    dir_cos = img.affine[:3, :3] / im_zooms
+
+    # Initialize the affine of the B-Spline grid
+    bs_affine = np.eye(4)
+    bs_affine[:3, :3] = np.array(control_zooms_mm) * dir_cos
+    bs_zooms = nb.affines.voxel_sizes(bs_affine)
+
+    # Calculate the shape of the B-Spline grid
+    im_extent = im_zooms * (im_shape - 1)
+    bs_shape = (im_extent // bs_zooms + 3).astype(int)
+
+    # Center both images
+    im_center = img.affine @ np.hstack((0.5 * (im_shape - 1), 1))
+    bs_center = bs_affine @ np.hstack((0.5 * (bs_shape - 1), 1))
+    bs_affine[:3, 3] = im_center[:3] - bs_center[:3]
+
+    return nb.Nifti1Image(np.zeros(bs_shape, dtype="float32"), bs_affine)
+
+
+def bspline_weights(points, level):
+    """Calculate the tensor-product cubic B-Spline weights for a list of 3D points."""
+    ctl_spacings = [float(sp) for sp in level.header.get_zooms()[:3]]
+    ncoeff = level.dataobj.size
+    ctl_points = (
+        level.affine.astype("float32") @ (
+            np.vstack((
+                np.argwhere(np.asanyarray(level.dataobj) == 0.0).astype("float32").T,
+                np.ones((1, ncoeff), dtype="float32")
+            ))
+        )
+    )[:3].T
```

Comment:
This can be another apply_affines case.

Also, are you guaranteed to have exactly zeros? Do you want to use `isclose`?

```suggestion
    ctl_points = nb.affines.apply_affine(
        level.affine.astype("float32"),
        np.argwhere(np.isclose(level.dataobj, 0)))
```

Reply:
Should be zero because the image object is initialized with zeros before, but yes, this should be safer.